test/integration/thermal_camera.cc

```diff
@@ -448,6 +454,176 @@ void ThermalCameraTest::ThermalCameraBoxes8Bit(
+    thermalCamera->SetHFOV(hfov);
+    EXPECT_NEAR(thermalCamera->HFOV().Radian(), hfov, DOUBLE_TOL);
+
+    // set bit depth
+    thermalCamera->SetImageFormat(ignition::rendering::PF_L8);
+    EXPECT_EQ(ignition::rendering::PF_L8, thermalCamera->ImageFormat());
+
+    // set min max temp
+    thermalCamera->SetMinTemperature(minTemp);
+    EXPECT_DOUBLE_EQ(minTemp, thermalCamera->MinTemperature());
+    thermalCamera->SetMaxTemperature(maxTemp);
+    EXPECT_DOUBLE_EQ(maxTemp, thermalCamera->MaxTemperature());
+
+    // thermal-specific params
+    // set room temperature: 294 ~ 298 Kelvin
+    float ambientTemp = 296.0f;
+    float ambientTempRange = 4.0f;
+
+    // 8 bit format so higher number here (lower resolution)
+    // +- 3 degrees
+    float linearResolution = 3.0f;
+    thermalCamera->SetAmbientTemperature(ambientTemp);
+    EXPECT_FLOAT_EQ(ambientTemp, thermalCamera->AmbientTemperature());
+    thermalCamera->SetAmbientTemperatureRange(ambientTempRange);
+    EXPECT_FLOAT_EQ(ambientTempRange, thermalCamera->AmbientTemperatureRange());
+    thermalCamera->SetLinearResolution(linearResolution);
+    EXPECT_FLOAT_EQ(linearResolution, thermalCamera->LinearResolution());
+    scene->RootVisual()->AddChild(thermalCamera);
+
+    // Set a callback on the camera sensor to get a thermal camera frame
+    // todo(anyone) change this to uint8_t when thermal cameras supports a
+    // ConnectNewThermalFrame event that provides this format
+    uint16_t *thermalData = new uint16_t[imgHeight * imgWidth];
+    ignition::common::ConnectionPtr connection =
+      thermalCamera->ConnectNewThermalFrame(
+          std::bind(&::OnNewThermalFrame, thermalData,
+            std::placeholders::_1, std::placeholders::_2, std::placeholders::_3,
+            std::placeholders::_4, std::placeholders::_5));
+    EXPECT_NE(nullptr, connection);
+
+    // thermal image indices
+    int midWidth = static_cast<int>(thermalCamera->ImageWidth() * 0.5);
+    int midHeight = static_cast<int>(thermalCamera->ImageHeight() * 0.5);
+    int mid = midHeight * thermalCamera->ImageWidth() + midWidth -1;
+    int left = midHeight * thermalCamera->ImageWidth();
+    int right = (midHeight+1) * thermalCamera->ImageWidth() - 1;
+
+    // Update a few times to make sure the flow of particles do not affect
+    // the readings
+    for (unsigned int i = 0; i < 100u; ++i)
+    {
+      thermalCamera->Update();
+
+      // verify temperature
+      // Box should be in the middle of image and return box temp
+      // Left and right side of the image frame should be ambient temp
+      EXPECT_NEAR(ambientTemp, thermalData[left] * linearResolution,
+          ambientTempRange);
+      EXPECT_NEAR(ambientTemp, thermalData[right] * linearResolution,
+          ambientTempRange);
+      EXPECT_FLOAT_EQ(thermalData[right], thermalData[left]);
+      EXPECT_NEAR(boxTemp, thermalData[mid] * linearResolution,
+          linearResolution);
+    }
```

Comment:
Instead of checking the same 3 locations every loop iteration (`left`, `mid`, and `right`), would it be better to randomly sample different image coordinates each loop iteration to make sure that we are checking a larger portion of the image? I think that the difficulty with a random sampling approach is that we'd have to determine if the coordinates map to the box temperature or ambient temperature. If this is too hard to implement, we can leave this part of the test as-is.

Reply:
as mentioned above, I tweaked the particle distance and params to fill the whole camera view and so we don't need to do random sampling here
4ff85d3